components/function-controller/pkg/apis/serverless/v1alpha1/function_validation.go

```diff
@@ -0,0 +1,123 @@
+package v1alpha1
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/api/validation"
+	v1validation "k8s.io/apimachinery/pkg/apis/meta/v1/validation"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"knative.dev/pkg/apis"
+)
+
+var reservedEnvs = []string{
+	"K_REVISION",
+	"K_CONFIGURATION",
+	"K_SERVICE",
+	"FUNC_RUNTIME",
+	"FUNC_HANDLER",
+	"FUNC_TIMEOUT",
+	"FUNC_PORT",
+	"PORT",
+	"MOD_NAME",
+	"NODE_PATH",
+}
+
+func (fn *Function) Validate(_ context.Context) (errors *apis.FieldError) {
+	errors = fn.validateObjectMeta(errors, "metadata")
+
+	spec := fn.Spec
+	errors = spec.validateSource(errors, "spec.source")
+	errors = spec.validateDeps(errors, "spec.deps")
+	errors = spec.validateEnv(errors, "spec.env")
+	errors = spec.validateLabels(errors, "spec.labels")
+	errors = spec.validateReplicas(errors, "spec")
+	errors = spec.validateResources(errors, "spec.resources")
+
+	return errors
+}
+
+func (fn *Function) validateObjectMeta(apisError *apis.FieldError, fldPath string) *apis.FieldError {
+	nameFn := validation.ValidateNameFunc(validation.NameIsDNS1035Label)
+	fieldPath := field.NewPath(fldPath)
+	if errs := validation.ValidateObjectMeta(&fn.ObjectMeta, true, nameFn, fieldPath); errs != nil {
+		for _, err := range errs {
+			apisError = apisError.Also(apis.ErrInvalidValue(err.Error(), err.Field))
+		}
+	}
+	return apisError
+}
+
+func (spec *FunctionSpec) validateSource(apisError *apis.FieldError, fldPath string) *apis.FieldError {
+	if strings.TrimSpace(spec.Source) == "" {
+		apisError = apisError.Also(apis.ErrMissingField(fldPath))
+	}
+	return apisError
+}
+
+func (spec *FunctionSpec) validateDeps(apisError *apis.FieldError, fldPath string) *apis.FieldError {
+	if deps := strings.TrimSpace(spec.Deps); deps != "" && (deps[0] != '{' || deps[len(deps)-1] != '}') {
+		apisError = apisError.Also(apis.ErrInvalidValue("deps should start with '{' and end with '}'", fldPath))
+	}
+	return apisError
+}
```

Comment:
currently we do some kind of validation/**mutation** in function controller, which should be moved here + into defaulting function 

https://github.com/kyma-project/kyma/blob/master/components/function-controller/internal/controllers/serverless/build_resources.go#L201 basically please investigate usage of this function in fun-controller, and move it here if needed